runner/main.go

```diff
@@ -0,0 +1,225 @@
+	}
+	testDescriptor struct {
+		runner  *testRunner
+		test    testing.InternalTest
+		ran     int
+		failed  bool
+		flaky   bool
+		error   bool
+		skipped bool
+	}
+	internalTestResult struct {
+		ran        bool      // Test or benchmark (or one of its subtests) was executed.
+		failed     bool      // Test or benchmark has failed.
+		skipped    bool      // Test of benchmark has been skipped.
+		done       bool      // Test is finished and all subtests have completed.
+		finished   bool      // Test function has completed.
+		raceErrors int       // number of races detected during test
+		name       string    // Name of test or benchmark.
+		start      time.Time // Time test or benchmark started
+		duration   time.Duration
+	}
+)
+
+var runner *testRunner
+var runnerRegexName *regexp.Regexp
+
+// Runs a test suite
+func Run(m *testing.M, exitOnError bool, failRetriesCount int) int {
+	runner := &testRunner{
+		m:                m,
+		exitOnError:      exitOnError,
+		failRetriesCount: failRetriesCount,
+	}
+	runner.init()
+	return runner.Run()
+}
+
+// Initialize test runner, replace the internal test with an indirection
+func (r *testRunner) init() {
+	if tPointer, err := getFieldPointerOfM(r.m, "tests"); err == nil {
+		tests := make([]testing.InternalTest, 0)
+		internalTests := (*[]testing.InternalTest)(tPointer)
+		for _, test := range *internalTests {
+			td := &testDescriptor{
+				runner: r,
+				test:   test,
+				ran:    0,
+				failed: false,
+			}
+			tests = append(tests, testing.InternalTest{
+				Name: test.Name,
+				F:    td.run,
+			})
+		}
+		// Replace internal tests
+		*internalTests = tests
+	}
+}
+
+// Runs the test suite
+func (r *testRunner) Run() int {
+	return r.m.Run()
+}
+
+// Internal test runner, each test calls this method in order to handle retries and process exiting
+func (td *testDescriptor) run(t *testing.T) {
+	run := 1
+	maxRetries := td.runner.failRetriesCount
+	exitOnError := td.runner.exitOnError
+	var rc interface{}
+
+	for {
+		var innerTest *testing.T
+		title := "Run"
+		if run > 1 {
+			title = "Retry:" + strconv.Itoa(run-1)
+		}
+		title = "[runner." + title + "]"
+		t.Run(title, func(it *testing.T) {
+			defer func() {
+				rc = recover()
+				if rc != nil {
+					it.FailNow()
+				}
+			}()
+			innerTest = it
+			td.test.F(it)
+		})
+		innerTestInfo := getTestResultsInfo(innerTest)
+
+		if rc != nil {
+			if exitOnError {
+				panic(rc)
+			}
+			fmt.Println("PANIC RECOVER:", rc)
+			td.error = true
+		}
+		td.skipped = innerTestInfo.skipped
+		td.ran++
+
+		// Current run failure
+		if innerTestInfo.failed {
+			td.failed = true
+		}
+		// Current run ok but previous with fail -> Flaky
+		if !innerTestInfo.failed && td.failed {
+			td.failed = false
+			td.flaky = true
+			break
+		}
+
+		if td.skipped || !td.failed || run > maxRetries {
+			break
+		}
+		run++
+	}
+	if td.flaky {
+		fmt.Println("*** FLAKY", t.Name())
+	}
+	if td.error && exitOnError {
+		panic(rc)
+	}
```

Comment:
It will. I've always recover from panic in the subtest for flaky test handling, so if the `exitOnError` flag is set, `rc` is the original panic info from the subtest.